File: modules/SeedSize/SeedSize.pl
#!/usr/bin/perl
package BQPhytomorph::SeedSize;

use warnings;
use strict;

use Cwd 'abs_path';
use File::Basename;
use File::Copy;
use File::Glob;
use Getopt::Long;
use XML::LibXML;
use XML::Simple;
use Text::CSV;
use POSIX qw(strftime);
use URI::Escape;
use Log::Log4perl;

use BQPhytomorph;

our $EXEC          = "./seedSize";

my $log = Log::Log4perl::get_logger('seedsize');

#########################################
## Argument handling
my ($help, $image, $mex, $user, $password, $userpass, $staging_path, $token);
my $series_id = '';
my $debug = 0;
my $dryrun = 0;

usage() if ( ! GetOptions('help|?' => \$help, 
                          'debug|d' => \$debug,
                          'dryrun|n' => \$dryrun,
                          'mex_url=s' => \$mex,
                          'image_url=s' => \$image, 
                          'staging_path=s'=>\$staging_path,
                          'user=s' => \$user,
                          'password=s' => \$password,
                          'token=s' => \$token,
)
          or defined $help );

my $options = { debug=> $debug, dryrun => $dryrun };
my $command  = shift;

$command = 'start' if (!defined $command);

if ($user && $password) {
  $userpass = $user . ':' . $password;
}


print "command is " . $command . "\n";

my $ret=0;
$ret = command_setup() if ($command eq 'setup');
$ret = command_teardown() if ($command eq 'teardown');
$ret = command_start() if ($command eq 'start');

exit($ret);


sub usage {
  print "Unknown option: @_\n" if ( @_ );
  print "usage: SeedSize [--mex_url=URL] [--user=USER --password=PASS|token=TOKEN] [--image_url=IMAGE] [--staging_path=SERIES_ID] [--debug] [--help]  [start|finished|test|test_finished] \n";
  exit 1;
}


sub command_setup {
  my $bq = new BQPhytomorph(mex_url => $mex, userpass => $userpass,
                            token=> $token, options=>$options );
  my $imgdir = "$staging_path/images";
  unless ( -d $imgdir ) {
    mkdir $imgdir or die "Can't make  $imgdir";
  }

  $bq->update_mex('initializing');


  my @localfiles;
  if ($image =~ m/dataset/g) {
    @localfiles = $bq->fetch_dataset ($image,  $imgdir);
  } else {
    @localfiles = $bq->fetch_image ($image,  $imgdir);
  }

  $bq->update_mex('scheduling');
  return 0;
}
sub command_start {
  my $bq = new BQPhytomorph(mex_url => $mex, userpass => $userpass,
                            token=> $token, options=>$options );

  $bq->update_mex('running');

  my $imgdir = "$staging_path/images/";
  return system ($EXEC, $imgdir);
}

sub command_teardown {
  my $bq = new BQPhytomorph(mex_url => $mex, userpass=>$userpass,
                            token=>$token,
                            options=>$options) ;
  # Collect results, process and save

  my @localfiles = glob "$staging_path/images/*C.csv";
  my $summary    = "$staging_path/images/summary.csv";

  my @mextags;
  eval {
    for my $mc (@localfiles) {
      if (-f $mc) {
        push (@mextags, saveStats ($bq, $mex, $image, $mc ) )
      }
    }
    1;
  } or do {
    $log->error("problems while saving data $@");
    print "FAILED";
    $bq->finish_mex ( status => 'FAILED', msg=> $@ );
    return 1;
  };
  # Update the MEX with state
  print "FINISHED";
  $bq->finish_mex ( status => 'FINISHED', tags=> \@mextags );
  return 0;
}

sub saveStats {
  my ($bq, $mex, $image,  $MC) = @_;

  print "READING $MC\n";

  #area-0, minor axis len-1, major axis len-2, centerX-3, centerY-4, rotation-5, minorX-6, minorY-7, majorX-8,majorY-9

  my $csv = Text::CSV->new();
  my @seeds;
  open (MC, "<", $MC) or die "No $MC";
  while (<MC>) {
    if ($csv->parse($_)) {
      my @cols = $csv->fields();
      push ( @seeds,
        { type => "seed",
          tag => [ { name=>"area", value=> $cols[0] },
                   { name=>"major", value=> $cols[2] },
                   { name=>"minor", value=> $cols[1] } ],
          ellipse => { 
               vertex=> [ { x=> $cols[3], y=> $cols[4], index=>0 },
                          { x=> $cols[8], y=> $cols[9], index=>1 },
                          { x=> $cols[6], y=> $cols[7], index=>2 }
                        ]
                     }
        });
    }
  }
  close(MC);
  my $xs = new XML::Simple(RootName=>undef);


  # Add Gobjects directly to image
  my $gobreq = { gobject => { name=>"SeedSize",
                              gobject => \@seeds
                            }
               };
  my $content =  $xs->XMLout($gobreq);
  my $gurl    = "";
  my $doc;
  if ( ! $dryrun ) {
    $doc = $bq->postxml ($image . "/gobjects", $content);
    die "post failed:" . $image . "/gobjects" unless defined $doc;
    $gurl = $doc->findnodes('//gobject/@uri')->to_literal->value;
  }
  if ( $debug ) {
    print "GOB $gurl\n";
    print "GOB => $content";
  }

  # Add Tag to image also
  my $tagreq = { tag =>
                 { name => "SeedSize",
                   tag => [{ name=>"gobjects_url", type=>"link", value=>$gurl},
                           { name=>"mex_url", type=>"link", value=>$mex},
                           { name=>'date_time', value=> strftime "%Y-%m-%d %H:%M:%S", localtime },
                           { name=>"area-histogram", type=>"statistics",
                             value => ("http:/stats?url=" . uri_escape($gurl) 
                                       ."&xpath=".uri_escape('//tag[@name="area"]')
                                       ."&xmap=".uri_escape("tag-value-number")
                                       ."&xreduce=histogram&run=true")},

                           { name=>"major-histogram", type=>"statistics",
                             value => ("http:/stats?url=" . uri_escape($gurl) 
                                       ."&xpath=".uri_escape('//tag[@name="major"]')
                                       ."&xmap=".uri_escape("tag-value-number")
                                       ."&xreduce=histogram&run=true")},

                           { name=>"minor-histogram", type=>"statistics",
                             value => ("http:/stats?url=" . uri_escape($gurl) 
                                       ."&xpath=".uri_escape('//tag[@name="minor"]')
                                       ."&xmap=".uri_escape("tag-value-number")
                                       ."&xreduce=histogram&run=true")},
                           { name=>"csv-area-major-minor", type=>"file",
                             value => ("http:/stats/csv?url=" . uri_escape($gurl)
                                       ."&xpath=".uri_escape('//tag[@name="area"]')
                                       ."&xpath1=".uri_escape('//tag[@name="major"]')
                                       ."&xpath2=".uri_escape('//tag[@name="minor"]')
                                       ."&xmap=".uri_escape("tag-value-number")
                                       ."&xreduce=vector")},
                          ]
                 }
               };
  $content =  $xs->XMLout($tagreq);
  if (! $dryrun ) {
    $doc = $bq->postxml ($image . "/tags", $content);
    die "post failed:" . $image . "/tags" unless defined $doc;
  }
  if ( $debug ) {
    print "TAG => $content";
  }
  # Add the pointers in the mex

  my $mextag = {name=>"gobjects_url", type=>"link", value=>$gurl};

  return $mextag
}

File: modules/SeedSize/SeedSize.py
#!/usr/bin/env python
import os
import sys
import optparse 
import subprocess
import glob
import csv
import pickle
import logging
import itertools

from bq.api import BQSession
from bq.api.util import fetch_dataset, fetch_image_pixels, d2xml



EXEC = "./seedSize"
IMAGE_MAP = "image_map.txt"

def gettag (el, tagname):
    for kid in el:
        if kid.get ('name') == tagname:
            return kid, kid.get('value')
    return None,None
            
class SeedSize(object):

    def setup(self):
        if not os.path.exists(self.images):
            os.makedirs(self.images)

        self.bq.update_mex('initializing')
        if self.is_dataset:
            results = fetch_dataset(self.bq, self.resource_url, self.images, True)
        else:
            results = fetch_image_pixels(self.bq, self.resource_url, self.images, True) 

        with open(self.image_map_name, 'wb') as f:
            pickle.dump(results, f)
        return 0
        

    def start(self):
        self.bq.update_mex('executing')
        # Matlab requires trailing slash
        r = subprocess.call([EXEC, 'images/'])
        return r

    def teardown(self):
        with  open(self.image_map_name, 'rb') as f:
            self.url2file = pickle.load(f) # 
            self.file2url =  dict((v,k) for k,v in self.url2file.iteritems())

        summary    = os.path.join(self.images, 'summary.csv')

        summary_tags = self._read_summary(summary)

        # Post all submex for files and return xml list of results
        tags = []
        gobjects = []
        if not self.is_dataset:
            localfiles = glob.glob(os.path.join(self.images, '*C.csv'))
            gobs = self._read_results(localfiles[0])
            gobjects = [ { 'name': 'SeedSize', 'gobject':gobs } ]
            #tags = [{ 'name':'image_url', 'value' : self.resource_url}]
            tags = summary_tags
        else:
            mexlist = self._post_submex()
            tags.extend(summary_tags)
            for i, submex in enumerate(mexlist):
                tag, image_url = gettag(submex, 'image_url')
                gob, gob_url = gettag(submex, 'SeedSize')

                mexlink = { 'name' : 'submex', 
                            'tag'  : [{ 'name':'mex_url', 'value':submex.get('uri')},
                                      { 'name':'image_url', 'value' : image_url},
                                      { 'name':'gobject_url', 'value' : gob.get('uri') } ]
                            }
                tags.append(mexlink)

        self.bq.finish_mex(tags = tags, gobjects = gobjects)
        return 0


    def _post_submex(self):
        submex = []

        localfiles = glob.glob(os.path.join(self.images, '*C.csv'))
        result2url = dict( (os.path.splitext(f)[0] + 'C.csv', u) for f, u in self.file2url.items())
        

        for result in localfiles:
            gobs = self._read_results(result)
            if result not in result2url:
                log.error ("Can't find url for %s given files %s and map %s" % 
                           result, localfile, result2url)
            mex = { 'type' : self.bq.mex.type,
                    'name' : self.bq.mex.name,
                    'value': 'FINISHED', 
                    'tag': [ {'name': 'image_url', 'value':  result2url [result]},
                             {'name': 'resource_url', 'value': self.config.resource_url }],
                    'gobject' : { 'name' : 'SeedSize', 'gobject' : gobs },
                    
                    }
            submex.append (mex)
        
        url = self.bq.service_url('data_service', 'mex', query={ 'view' : 'deep' })
        response = self.bq.postxml(url, d2xml({'request' : {'mex': submex}} ))

        return response
        

    def _read_summary(self, csvfile):
        #%mean(area), mean(minoraxislen), mean(majoraxislen), standarddev(area),
        #standarddev(minoraxislen), standarddev(majoraxislen), total seedcount,
        #mean thresholdused, weighted mean of percentclusters1, weighted mean of percentclusters2
        f= open(csvfile,'rb')
        rows = csv.reader (f)
        tag_names = [ 'mean_area', 'mean_minoraxis', 'mean_majoraxis',
                      'std_area', 'std_minoraxis', 'std_majoraxis',
                      'seedcount',
                      'mean_threshhold', 
                      'weighted_mean_cluster_1','weighted_mean_cluster_2',
                      ]

        # Read one row(rows.next()) and zip ( name, col) unpacking in d2xml format
        summary_tags = [ { 'name': n[0], 'value' : n[1] } 
                         for n in itertools.izip(tag_names, rows.next()) ] 
        f.close()
        
        return  [ {'name':'summary', 'tag':summary_tags } ]

    def _read_results(self, csvfile):
        results  = []
        f= open(csvfile,'rb')
        rows = csv.reader (f)
        for col in rows:
            results.append( {
                    'type' : 'seed',
                    'tag' : [ { 'name': 'area', 'value': col[0]},
                              { 'name': 'major', 'value': col[2]},
                              { 'name': 'minor', 'value': col[1]} ],
                    'ellipse' : { 
                        'vertex' : [ { 'x': col[3], 'y':col[4], 'index':0 },
                                     { 'x': col[8], 'y':col[9], 'index':1 },
                                     { 'x': col[6], 'y':col[7], 'index':2 }]
                        }
                    })
        f.close()
        return results



    def run(self):
        parser  = optparse.OptionParser()
        parser.add_option('-d','--debug', action="store_true")
        parser.add_option('-n','--dryrun', action="store_true")
        parser.add_option('--resource_url')
        parser.add_option('--mex_url')
        parser.add_option('--staging_path')
        parser.add_option('--auth_token')
        parser.add_option('--credentials')

        (options, args) = parser.parse_args()


        if options.auth_token:
            self.bq = BQSession().init_mex(options.mex_url, options.auth_token)
        else:
            user,pwd = options.credentials.split(':')
            self.bq = BQSession().init_local(user,pwd)

        if len(args) != 1 or options.resource_url is None:
            parser.error('Need a command and resource_url')

        command = args.pop(0)

        if command not in ('setup','teardown', 'start'):
            parser.error('Command must be start, setup or teardown')

        logging.basicConfig(level=logging.DEBUG)

        # maltab code requires trailing slash..
        self.images = os.path.join(options.staging_path, 'images') + os.sep
        self.image_map_name = os.path.join(options.staging_path, IMAGE_MAP)
        self.resource_url = options.resource_url
        self.config = options
        self.is_dataset = 'dataset' in self.resource_url



            
        command = getattr(self, command)
        try:
            r = command()
        except:
            logging.exception ("problem during %s" % command)
            sys.exit(1)
        
        sys.exit(r)



if __name__ == "__main__":
    SeedSize().run()
    
